### ISSUE TITLE
ImageMagick 6.7.6.4 produces all black images due to -colorspace RGB argument

### ISSUE THREAD

**Issue by Dazzel** [flourishlib/flourish-classes#38]
Ticket #748 by @audvare

Line 1182 in fImage.php 

``` php
if (strpos($command_line, ' -colorspace ') === FALSE) {
            $command_line .= ' -colorspace RGB ';
        }
```

Without this argument, the image converts fine.

As a temporary solution, I downgraded to ImageMagick? 6.7.5.3. There might be an issue with this version: http://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=20682

**Comment by Cruel** [comment]
This is a big problem for me as well (ImageMagick 6.7.8-9). With -colorspace RGB it darkens all my images, but without it it works fine.
Here's an original image: http://i.imgur.com/hJqqj.jpg
And here's resize() called on it: http://i.imgur.com/xiwvD.jpg

I just removed the default colorspace lines, hopefully there's no unintended issues with that, but I guess I'll find out.

**Comment by Tatsh** [comment]
Now the issue is with -colorspace RGB the colours are way too dark.

Original: http://ky.tatsh.net/files/images/img_0246_copy1.jpg
Resized to 250px width: http://ky.tatsh.net/files/images/thumbs/img_0246.jpg

Command line:

```
'/usr/bin/convert' \
'jpg:/home/tatsh/dev/kusaba-y/files/images/thumbs/__flourish_temp/img_0246.jpg[0]' \
-resize "250x187" \
-colorspace RGB \
-compress JPEG \
-quality 90 \
'jpg:/home/tatsh/dev/kusaba-y/files/images/thumbs/__flourish_temp/img_0246.jpg' 2>&1
```

**Comment by wbond** [comment]
It seems there are two options:
1. Downgrade to a functioning version of ImageMagick
2. Change fImage to detect the color palette of the original file and convert to RGB if the original is CMYK and the destination is JPG or PNG. This flag was originally added to handle resizing CMYK tiff files to JPGs since my last company wrote an app that ran into that situation quite frequently.

**Comment by Tatsh** [comment]
I would prefer option 2 but I am unsure it is necessary, and for processing many images at once, would that not be a performance hindrance?

My solution is to just remove defaulting to any colorspace and let ImageMagick do the work. This means removing the check and adding. I already did this in my fork and I am having no problems (tested mostly pictures that come off my iPhone which uses sRGB). A user on the IRC channel for ImageMagick suspects a gamma conversion is occurring, but is unsure.

https://github.com/Tatsh/flourish-classes/commit/e2a9f262b093acc7e41b6798537b55cddf12ce6a

**Comment by wbond** [comment]
This issue with removing this is that resulting JPGs end up being CMYK JPGs, which don't display at all in some browsers.

**Comment by Tatsh** [comment]
That is definitely worrying. Even IE 8 cannot show CMYK images.

The only way to check easily I think is to read using `identify -verbose` and get the line that has `jpeg:colorspace: n` (parse it) and fallback to `exif_read_data()`. Sometimes photos have no EXIF data so in the worse the case `-colorspace` argument may have to default to RGB.

**Comment by Tatsh** [comment]
According to this page, ImageMagick 6.7.5 and up set the default colorspace to sRGB.

http://www.imagemagick.org/Usage/resize/#resize_colorspace

**Comment by wbond** [comment]
Yeah, according to that page, it appears previous to 6.7.5, RGB and sRGB were swapped. To it appears the "fix" is to detect the version of ImageMagick and change the flag accordingly.

**Comment by Tatsh** [comment]
This should come close and works for me.

https://github.com/Tatsh/flourish-classes/commit/3b112c57ce5e901dce3e3b59f9e09e176490dc3d
https://github.com/Tatsh/flourish-classes/commit/492bbe2351c9ee282ba8490ffc19f3ebe8975a21
